cql3/functions/user_aggregate.hh

```diff
@@ -19,6 +19,8 @@ namespace functions {
 class user_aggregate : public db::functions::aggregate_function, public data_dictionary::keyspace_element {
 public:
     user_aggregate(function_name fname, bytes_opt initcond, ::shared_ptr<scalar_function> sfunc, ::shared_ptr<scalar_function> reducefunc, ::shared_ptr<scalar_function> finalfunc);
+    virtual ~user_aggregate() override = default;
```

Comment:
> may lead to calling the wrong destructor or memory leaks

yes, but just "may". as you've found out, we've been using classes with virtual methods but without virtual destructors in some places in scylla. i think it's fine. as long as we do not destruct an instance of the derived class with an pointer pointing to its parent class.

have you identified any issue because of this? i just wanted to point out that marking the dtor `virtual` is not free. then we need to pay for the overhead of looking up the function pointer in the vtable when destructing the derived class instance even we are aware of the actual type of instance, and we are destructing the instance with that type.

Reply:
> have you identified any issue because of this?

No. The motivation for this PR was that I noticed we use `shared_ptr<const keyspace_element>`s in code. I'm honestly surprised ASAN doesn't report any leaks because although `~keyspace_element()` is a no-op, those pointers don't seem to point to any pre-existing objects that would call the right destructor. Maybe `seastar::shared_ptr` somehow goes through the vtable anyway? Unfortunately, I don't feel competent enough to try to explain it.

In any case, even if the code is correct (and it seems to be; I doubt ASAN has a bug like that), the type itself is supposed to describe an element of a keyspace and there shouldn't be that many of them, so the vtable overhead (as far as I can tell at the moment) should not be a big problem (note: it may change after my PR – scylladb/scylladb#20168 – is merged because we decided to abuse it a bit, e.g. [here](https://github.com/scylladb/scylladb/pull/20168/files#diff-0969c0a6117da3f166a706f7f8301edac520a64e96cfed57cc337e92eabfaa74R88)).

At the same time, I understand if this PR is rejected

Reply:
> > have you identified any issue because of this?
> 
> No. The motivation for this PR was that I noticed we use `shared_ptr<const keyspace_element>`s in code. I'm honestly surprised ASAN doesn't report any leaks because although `~keyspace_element()` is a no-op, those pointers don't seem to point to any pre-existing objects that would call the right destructor. Maybe `seastar::shared_ptr` somehow goes through the vtable anyway? Unfortunately, I don't feel competent enough to try to explain it.

regarding to why we survive ASan's check. this is an interesting topic, and it applies to `std::shared_ptr` as well. 
let me elaborate this a little bit. like `std::shared_ptr` , `seastar::shared_ptr` creates a type erased deleter object when it is created. in the case of `seastar::shared_ptr`, the deleter is of type `shared_ptr_count_base`, and each `shared_ptr` holds a pointer pointing to a concrete class derived from `shared_ptr_count_base`, and the concrete class understands how to destroy the pointee of the smart pointer.

and what we typical do is like
```c++
#include <memory>

class base {
    int value;
};
class derived : public base {
   int another_value;
};

int main() {
    shared_ptr<base> b = make_shared<derived>();
}
```

when we assign the return value of `make_shared<derived>()` to `b`, we call into the constructor whose signature looks like:
```c++
template <typename U, typename = std::enable_if_t<std::is_base_of_v<T, U>>>
    shared_ptr(shared_ptr<U>&& x) noexcept
            : _b(x._b)
            , _p(x._p) {
        x._b = nullptr;
        x._p = nullptr;
    }
```
so, the lhs of the assignment expression gets the type erased deleter from the rhs of the expression. in other words, when `b` is destroyed, if the deleter's ref count reaches zero,  the original deleter gets called. and the proper destructor is used. if you take a look at `seastar::shared_ptr::~shared_ptr()` , you'd notice that we don't call `delete _p`.

the same applies to assignment operator of `shared_ptr` -- the type erased deleter is either transferred to the new owner, or assigned to the new owner.

> 
> In any case, even if the code is correct (and it seems to be; I doubt ASAN has a bug like that), the type itself is supposed to describe an element of a keyspace and there shouldn't be that many of them, so the vtable overhead (as far as I can tell at the moment) should not be a big problem (note: it may change after my PR – #20168 – is merged because we decided to abuse it a bit, e.g. [here](https://github.com/scylladb/scylladb/pull/20168/files#diff-0969c0a6117da3f166a706f7f8301edac520a64e96cfed57cc337e92eabfaa74R88)).
> 

ASan is able to identify unpaired new and delete. at least in this case, ASan does not manifest a bug. our usage is sane. unless we allow constructing an instance of `seastar::shared_ptr<foo>` directly from `derived*`, we should be fine.

> At the same time, I understand if this PR is rejected

based on the analysis above, i would not approve this PR.  but i understand that that a class with virtual function but without virtual dtor looks suspicious, so i would not reject it either. i am just not convinced that it's a bug, or it's a pure win.

Reply:
Thank you for your explanation! It makes sense and I understand it now. In that case, I can close the PR